# Ejercicio 5
Alumna: Blanca Cano Camarero  
Correo: blancacano@correo.ugr.es

Dado un número $n$

In [19]:
n = 55883687
factor(n)

7393 * 7559

### Apartado primero 

Factoriza $n$ aplicando el método de la $\rho$ de Polard. ¿Cuántas iteraciones necesitas?
Sea $p1$ el mayor de sus factores primos y $p2$ el siguiente primo.

In [20]:
# Código auxiliar de prácticas ateriores
def Verbose_Rho_de_Polard(n:Integer, f) -> Integer:
    '''Devuelve un factor de p
    o n si es un fracaso
    '''
    mf = lambda x: f(x)% n
    #print(f"Cálculo de Rho de Polar para n={n}")
    x = 1
    y = 1
    d = 1
    paso = 0
    #print(f"Paso {paso}   x={x}    y={y}")
    while d == 1:
        paso += 1
        x = mf(x)
        y = mf(mf(y))
        d = gcd(abs(x-y),n)
        #print(f"Paso {paso}    x={x}    y={y}    mcd={d}")
    return d
f = lambda x : x^2 +1 

## Versión de sacar todos los factores de manera manual 
def SacaFactores(numero_factores, n, f ):
    factores = []
    m = n
    hay_factores = True 
    while numero_factores >= 0 and hay_factores:
        numero_factores -= 1
        d = Verbose_Rho_de_Polard(m,f)
        if d != m:
            factores.append(d)
            m = m // d
            print(f"Se tiene que n = multiplicar los elementos de {factores} y {m}\n")
        else:
            factores.append(m)
            print('='*60)
            hay_factores = False
    return factores

factores = SacaFactores(2, n,f )
print(f"Los factores de {n-1} son {factores}")

def Test_Lucas_Lehmer(n, bases):
    m = n-1
    for a in bases:
        # cumple condición 1
        if mod(a^m, n) == 1:
            condicion_dos = True
            for i in range (2,m):
                # si es divisor 
                if m % i == 0:
                    if (int(a**(m/i))% n) == 1:
                        condicion_dos =  False
                        break
            if condicion_dos == True:
                return a
    return 0

for p in factores:
    resultado = Test_Lucas_Lehmer(p,[2,3,5,7,11,13])
    mensaje = (
        f"p={p} es primo por el test Lucas-Lehmer con a = {resultado}",
        "No se ha encontrado un entero que satisfaga el test de Lucas-Lehmer para p = {p}"
    )[resultado == 0] 
    print(mensaje)

Se tiene que n = multiplicar los elementos de [7559] y 7393

Los factores de 55883686 son [7559, 7393]
p=7559 es primo por el test Lucas-Lehmer con a = 13
p=7393 es primo por el test Lucas-Lehmer con a = 5


A la vista de los resultado, tenemos que 
$p1 = 7559$ y $p2=7393$
### Apartado segundo   

Calcula las partes enteras de $\sqrt{p1}$ y $\sqrt{p2}$ con el algoritmo entero.

In [39]:
def VerboseParteEnteraRaizCuadrada(n):
    # Incicializamos a 
    a = n
    if is_even(n):
        a //= 2
    else:
        a = (a+1)//2
    menor = True 
    i = 1
    while menor:
        numerador = a**2 + n
        cociente =  numerador//(2*a)
        resto = numerador % (2*a)
        print(f"i:{i}  a_i:{a}  a^2 +n: {numerador} = {cociente} * {2*a} + {resto}")
        i += 1
        if cociente < a:
            a = cociente
        else:
            menor = False 
    print( "\n",f"La parte entera de  sqr({n}) es {a}", '\n','='*50,'\n')   
    return a 

for p in factores:
    VerboseParteEnteraRaizCuadrada(p)


def ParteEnteraRaizCuadrada(n):
    # Incicializamos a 
    a = n
    if is_even(n):
        a //= 2
    else:
        a = (a+1)//2
    menor = True 
    i = 1
    while menor:
        numerador = a**2 + n
        cociente =  numerador//(2*a)
        resto = numerador % (2*a)
        i += 1
        if cociente < a:
            a = cociente
        else:
            menor = False 
    return a 
    

i:1  a_i:3780  a^2 +n: 14295959 = 1890 * 7560 + 7559
i:2  a_i:1890  a^2 +n: 3579659 = 946 * 3780 + 3779
i:3  a_i:946  a^2 +n: 902475 = 476 * 1892 + 1883
i:4  a_i:476  a^2 +n: 234135 = 245 * 952 + 895
i:5  a_i:245  a^2 +n: 67584 = 137 * 490 + 454
i:6  a_i:137  a^2 +n: 26328 = 96 * 274 + 24
i:7  a_i:96  a^2 +n: 16775 = 87 * 192 + 71
i:8  a_i:87  a^2 +n: 15128 = 86 * 174 + 164
i:9  a_i:86  a^2 +n: 14955 = 86 * 172 + 163

 La parte entera de  sqr(7559) es 86 

86.94250974063263
i:1  a_i:3697  a^2 +n: 13675202 = 1849 * 7394 + 3696
i:2  a_i:1849  a^2 +n: 3426194 = 926 * 3698 + 1846
i:3  a_i:926  a^2 +n: 864869 = 466 * 1852 + 1837
i:4  a_i:466  a^2 +n: 224549 = 240 * 932 + 869
i:5  a_i:240  a^2 +n: 64993 = 135 * 480 + 193
i:6  a_i:135  a^2 +n: 25618 = 94 * 270 + 238
i:7  a_i:94  a^2 +n: 16229 = 86 * 188 + 61
i:8  a_i:86  a^2 +n: 14789 = 85 * 172 + 169
i:9  a_i:85  a^2 +n: 14618 = 85 * 170 + 168

 La parte entera de  sqr(7393) es 85 

85.98255637046388


### Apartado tercero 
Calcula la FCS de $\sqrt{p_1}$ y $\sqrt{p_2}$ con el algoritmo entero.

In [59]:
# Algoritmo 

def FraccionContinuaRaiz(p):
    parte_entera_raiz =  ParteEnteraRaizCuadrada(p)
    q_a = parte_entera_raiz
    FCS = [q_a]
    P_a = 0
    Q_a = 1
     
    for i in range(2*p-1):
        P_s = q_a*Q_a -P_a
        Q_s = (p - P_s*P_s)//Q_a
        q_s = (P_s + parte_entera_raiz)//Q_s
        FCS.append(q_s)
        P_a = P_s
        q_a = q_s
        Q_a = Q_s
    return FCS

def calcularPeriodo(ristra):
    ''' con esto calcuaremos el periodo de cada FCS
    Se fundamenta en que sabemos que son simétricos y que 
    acaba en el doble que el primero.
    '''
    primero = ristra[0]
    ultimo = 2*primero
    encontrado = False
    start = 1
    while not encontrado and start < len(ristra):
        start = ristra.index(ultimo,start)
        # Condición de simetría
        if ristra[1:start] == (ristra[1:start])[::-1]:
            encontrado = True 
    return start

numeros_que_mostrar = 200
for p in factores:
    FCS = FraccionContinuaRaiz(p)
    periodo = calcularPeriodo(FCS)
    print(f"Los {periodo+1} primeros componentes de la FCS de {p} son:\n{FCS[:periodo+1]}\n")


Los 41 primeros componentes de la FCS de 7559 son:
[86, 1, 16, 2, 1, 1, 6, 2, 1, 3, 1, 8, 2, 1, 2, 1, 3, 1, 33, 1, 85, 1, 33, 1, 3, 1, 2, 1, 2, 8, 1, 3, 1, 2, 6, 1, 1, 2, 16, 1, 172]

Los 118 primeros componentes de la FCS de 7393 son:
[85, 1, 56, 3, 18, 1, 3, 2, 5, 1, 12, 2, 1, 1, 1, 1, 3, 1, 3, 1, 6, 2, 1, 2, 21, 8, 7, 24, 2, 2, 1, 7, 2, 9, 1, 1, 1, 4, 1, 2, 1, 1, 4, 1, 1, 1, 2, 1, 14, 1, 9, 1, 4, 3, 3, 3, 1, 2, 3, 3, 2, 1, 3, 3, 3, 4, 1, 9, 1, 14, 1, 2, 1, 1, 1, 4, 1, 1, 2, 1, 4, 1, 1, 1, 9, 2, 7, 1, 2, 2, 24, 7, 8, 21, 2, 1, 2, 6, 1, 3, 1, 3, 1, 1, 1, 1, 2, 12, 1, 5, 2, 3, 1, 18, 3, 56, 1, 170]



# Ejercicio 6
Alumna: Blanca Cano Camarero  
Correo: blancacano@correo.ugr.es


Sea $p$ el factor del ejercicio 5 con mayor periodo, en mi caso $p=7393.$

In [61]:
p = factores[1]
print(f"p = {p}")

FCS_p = FraccionContinuaRaiz(p)
periodo = calcularPeriodo(FCS)
FCS_p = FCS_p[:periodo+1]
print(FCS_p)

p = 7393
[85, 1, 56, 3, 18, 1, 3, 2, 5, 1, 12, 2, 1, 1, 1, 1, 3, 1, 3, 1, 6, 2, 1, 2, 21, 8, 7, 24, 2, 2, 1, 7, 2, 9, 1, 1, 1, 4, 1, 2, 1, 1, 4, 1, 1, 1, 2, 1, 14, 1, 9, 1, 4, 3, 3, 3, 1, 2, 3, 3, 2, 1, 3, 3, 3, 4, 1, 9, 1, 14, 1, 2, 1, 1, 1, 4, 1, 1, 2, 1, 4, 1, 1, 1, 9, 2, 7, 1, 2, 2, 24, 7, 8, 21, 2, 1, 2, 6, 1, 3, 1, 3, 1, 1, 1, 1, 2, 12, 1, 5, 2, 3, 1, 18, 3, 56, 1, 170]


### Apartado primero 
Calcula los convergentes de $\sqrt p$

In [77]:
# Calculamos los convergentes 

def Convergentes(FCS):
    A_aa = 1
    A_a = FCS[0]
    B_aa = 0
    B_a = 1
    convergentes = [A_a/B_a]
    for q in FCS[1:]:
        A_s = q* A_a + A_aa
        B_s = q*B_a + B_aa
        convergentes.append(A_s /B_s)

        #actualizamos pesos
        A_aa, A_a = A_a, A_s
        B_aa, B_a = B_a, B_s
    return convergentes

print(f"Los convergentes de sqrt({p}) son:")
Convergentes(FCS_p)

Los convergentes de sqrt(7393) son:


[85,
 86,
 4901/57,
 14789/172,
 271103/3153,
 285892/3325,
 1128779/13128,
 2543450/29581,
 13846029/161033,
 16389479/190614,
 210519777/2448401,
 437429033/5087416,
 647948810/7535817,
 1085377843/12623233,
 1733326653/20159050,
 2818704496/32782283,
 10189440141/118505899,
 13008144637/151288182,
 49213874052/572370445,
 62222018689/723658627,
 422545986186/4914322207,
 907313991061/10552303041,
 1329859977247/15466625248,
 3567033945555/41485553537,
 76237572833902/886663249525,
 613467616616771/7134791549737,
 4370510889151299/50830204097684,
 105505728956247947/1227059689894153,
 215381968801647193/2504949583885990,
 536269666559542333/6236958857666133,
 751651635361189526/8741908441552123,
 5797831114087869015/67430317948530994,
 12347313863536927556/143602544338614111,
 116923655885920217019/1359853216996057993,
 129270969749457144575/1503455761334672104,
 246194625635377361594/2863308978330730097,
 375465595384834506169/4366764739665402201,
 1748057007174715386270/20330367936